core/modules/rproc/InfileMerger.cc

```diff
@@ -214,19 +215,27 @@ bool InfileMerger::merge(std::shared_ptr<proto::WorkerResponse> response) {
     LOGS(_log, LOG_LVL_DEBUG, queryIdJobStr << " mergeDur=" << mergeDur.count());
     /// Check the size of the result table.
     if (_sizeCheckRowCount >= _checkSizeEveryXRows) {
+        auto tSize = _getResultTableSizeMB();
         LOGS(_log, LOG_LVL_DEBUG, queryIdJobStr << "checking ResultTableSize " << _mergeTable
-                                  << " " << _sizeCheckRowCount
+                                  << " " << tSize
                                   << " max=" << _maxResultTableSizeMB);
         _sizeCheckRowCount = 0;
-        auto tSize = _getResultTableSizeMB();
         if (tSize > _maxResultTableSizeMB) {
-            // TODO:DM-11524 Try deleting invalid rows if there are any, then check size again
-            std::ostringstream os;
-            os << queryIdJobStr << " cancelling queryResult table " << _mergeTable
-               << " too large at " << tSize << "MB max allowed=" << _maxResultTableSizeMB;
-            LOGS(_log, LOG_LVL_WARN, os.str());
-            _error = util::Error(-1, os.str(), -1);
-            return false;
+            // Try deleting invalid rows if there are any, then check size again
+            bool invalidResult = _invalidJobAttemptMgr.holdMergingForRowDelete("Checking size");
+            tSize = _getResultTableSizeMB();
+            if (tSize > _maxResultTableSizeMB || invalidResult) {
+                std::ostringstream os;
+                os << queryIdJobStr << " cancelling queryResult table " << _mergeTable;
+                if (invalidResult) {
+                    os << " failed to delete invalid rows.";
+                } else {
+                    os << " too large at " << tSize << "MB max allowed=" << _maxResultTableSizeMB;
+                }
+                LOGS(_log, LOG_LVL_ERROR, os.str());
+                _error = util::Error(-1, os.str(), -1);
+                return false;
+            }
```

Comment:
I'm thinking that we can probably avoid this entirely and only do cleanup when we return result. Presumably disk size is (almost) infinite or at least is much larger than a limit on a single result size. If we need to worry about free space on disk (and I think we will need to worry about that at some point when we implement result caching) we'll need a separate mechanism to keep result sizes under control, but it will probably take into account sizes of all results in cache.

Reply:
OK, thinking again with clear head after morning coffee - I think we really should get rid of cleanup at this point. Even if you remove rows from cancelled chunk you will have to re-run query on the same chunk anyways, and that will bring the same or larger number of rows again. I think it's safe to say that if partial result is larger than limit then final result will be larger still and just stop here, no cleanup needed.

Reply:
It's very possible that chunks in question have several invalid rows from failed attempts and all the rows from a complete query in the result table. The query doesn't need to be run again, as all the results are already there. So that would reduce the size of the result table, possibly significantly when hundreds of chunks are involved. In large results, it is easy for hundreds of chunks to be affected even if only one worker goes down.

The delete SQL query is only run when there are elements in _invalidJAWithRows, and that is cleared after the delete query since all the invalid rows should be gone.  So, if the delete query runs here, it wont run again at the end unless there are more failures.

Last, this really shouldn't run very often. It is only run if the result table is too large and the delete query will only run if it knows that there are invalid rows in the table.